Fix: Incorrect order of checks for TypeBuilder `GetConstructor` and `GetField` method

BartoszKlonowski opened the pull request:
This pull request fixes #45988

It changes the order of checks for `GetConstructor` and `GetField` method according to the workflow of `GetMethod`.

Full explanation can be found in the linked issue.

---

Because of reordered checks of those method, unit tests had to be adjusted.
For both `GetField` and `GetConstructor` they now check whether for not generic type the element will be created without throwing.

BartoszKlonowski committed:
Reorder checks for TypeBuilder.GetConstructor()

BartoszKlonowski committed:
Reorder checks for TypeBuilder.GetField()

BartoszKlonowski committed:
Adjust unit test to reordered TypeBuilder checks

janvorli commented:
cc: @eerhardt

eerhardt commented on src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderGetConstructor.cs:
```diff
@@ -7,17 +7,6 @@ namespace System.Reflection.Emit.Tests
     public class TypeBuilderGetConstructor
     {
-        [Fact]
-        public void GetConstructor_DeclaringTypeOfConstructorNotGenericTypeDefinition_ThrowsArgumentException()
```
We should not be removing these tests. Instead we should keep them, and if they no longer throw, we should assert what they do is correct.

BartoszKlonowski replied:
That's the case from the commit f2efe72, but locally this implementation of these tests did pass, but it was failing on CI checks for Helix.
Could you please that commit and let me know what could be done with that?

eerhardt replied:
I'm not sure I follow your question.

> but it was failing on CI checks for Helix.

Was it just failing for mono? Maybe we need to fix/change mono as well?

BartoszKlonowski replied:
@eerhardt Sure, now I see it's Mono that is failing.
Can you somehow (even briefly) guide me through how to handle the change for Mono?

eerhardt replied:
The Mono specific code is located here:

https://github.com/dotnet/runtime/blob/2ac5e2bb8a48dcb3038307338a04914ba708edc2/src/mono/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs#L1840-L1865

eerhardt commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs:
```diff
@@ -107,15 +107,15 @@ public static ConstructorInfo GetConstructor(Type type, ConstructorInfo construc
 
-            if (type.GetGenericTypeDefinition() != constructor.DeclaringType)
-                throw new ArgumentException(SR.Argument_InvalidConstructorDeclaringType, nameof(type));
+            if (!(type is TypeBuilderInstantiation))
```
While we are in here, we could re-write this to make it more modern:

```C#
            if (type is not TypeBuilderInstantiation typeBuilderInstantiation )
                throw new ArgumentException(SR.Argument_NeedNonGenericType, nameof(type));

            return ConstructorOnTypeBuilderInstantiation.GetConstructor(constructor, typeBuilderInstantiation);
```

BartoszKlonowski committed:
Use 'not' keyword for pattern matching

BartoszKlonowski committed:
Apply the same Type checks for Mono

GetConstructor and GetField had the same incorrect checks order
comparing to GetMethod as in the CoreCLR.
To keep the consistent solution, Mono has also been adjusted.

eerhardt commented:
@BartoszKlonowski - looks like it is still failing on mono.

BartoszKlonowski commented:
@eerhardt Unfortunately, I've tried to understand the order of checks for Mono - `GetMethod` seems to be a lot different than `GetField` and `GetConstructor`.
Can you guide me a bit through it? Otherwise I will need more time to analyse that.

eerhardt commented:
@lambdageek - any thoughts on the mono checks here? Can you help @BartoszKlonowski out?

lambdageek commented:
@BartoszKlonowski I think you need something like this change
https://github.com/dotnet/runtime/pull/53645/files#diff-1767a4830dd55e36974e0b34848e8f736f0f5e852866b7afc0abbe078f2d5dd1R133-R134

in Mono's GetField and GetConstructor - ie instead always throwing if the type is a generic type definition
https://github.com/dotnet/runtime/pull/53645/files#diff-621fff19cdf2db5ff0e021ff669fcd0d1bb6453d59d10b3769a2be8b624a565eR1924-R1925

you should instead check whether the generic type definition that `type` is instantiated from is the same as the declaring type of the `FieldInfo`.

BartoszKlonowski committed:
Check whether a declaring type isn't type's generic definition

BartoszKlonowski commented:
@lambdageek Thank you! Can you verify if I understood your advice correctly?

lambdageek commented on src/mono/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs:
```diff
@@ -1885,6 +1856,34 @@ private static bool IsValidGetMethodType(Type type)
+
+            if (type == null)
+                throw new ArgumentException("Type is not generic", nameof(type));
+
```
I think this needs the same code that `GetMethod` has:

```csharp
            if (type is TypeBuilder && type.ContainsGenericParameters)
                type = type.MakeGenericType(type.GetGenericArguments());
```

(and same for GetField)

BartoszKlonowski replied:
I've tried, but it won't build.

lambdageek replied:
Looks like you also need
```
[UnconditionalSuppressMessage("ReflectionAnalysis", "IL2055:UnrecognizedReflectionPattern",
            Justification = "Type.MakeGenericType is used to create a typical instantiation")]
```

BartoszKlonowski committed:
Make generic type if contains generic parameters

BartoszKlonowski committed:
Suppres UnrecognizedReflectionPattern for GetField and GetConstructor

eerhardt commented on src/mono/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs:
```diff
@@ -1835,35 +1835,6 @@ public GenericTypeParameterBuilder[] DefineGenericParameters(params string[] nam
-            return res;
-        }
-
         private static bool IsValidGetMethodType(Type type)
```
Can you put this method back where it was originally declared? That will make reviewing the change much easier. Also when people look at source history, will be able to see what change was made here easily.

eerhardt commented on src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderGetConstructor.cs:
```diff
@@ -8,14 +8,14 @@ namespace System.Reflection.Emit.Tests
     {
         [Fact]
-        public void GetConstructor_DeclaringTypeOfConstructorNotGenericTypeDefinition_ThrowsArgumentException()
+        public void GetConstructor_DeclaringTypeOfConstructorNotGenericTypeDefinition()
```
Can you make sure the cases where `ArgumentException` are thrown from the method are covered in tests? That way we know they are working correctly.

eerhardt replied:
I see a bunch of tests below have

```
[ActiveIssue("[https://github.com/dotnet/runtime/issues/2389](https://github.com/dotnet/runtime/issues/2389)", TestRuntimes.Mono)]
```

Are any of these fixed by your changes?

BartoszKlonowski replied:
I've just pushed all corrections (very sorry for the delays, I have very busy days recent times...)
Let's see which of the marked tests will pass and compare results against those reported in the mentioned issue.

eerhardt replied:
No need to apologize. We totally understand if people are busy.

Looks like the tests are now failing:

```
 Starting:    System.Reflection.Emit.Tests (parallel test collections = on, max threads = 2)
    System.Reflection.Emit.Tests.TypeBuilderGetConstructor.GetConstructor_DeclaringTypeOfConstructorWithoutGenericParameters_ThrowsArgumentException [FAIL]
      Assert.Equal() Failure
                 (pos 0)
      Expected: type
      Actual:   constructor
                 (pos 0)
      Stack Trace:
        /_/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs(123,0): at System.AssertExtensions.Throws[T](String expectedParamName, Func`1 testCode)
        /_/src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderGetConstructor.cs(16,0): at 
```

jeffhandley commented:
@eerhardt Since you're the most active reviewer on this, I assigned this PR to you for follow-up/decision before the RC1 snap.

eerhardt commented:
@BartoszKlonowski - I left 2 pieces of feedback above. Let me know if you will address these, or if you want me to resolve them. That way we can move this PR forward.

BartoszKlonowski committed:
Move GetMethodType validation method at original order

BartoszKlonowski committed:
Test GetConstructor against throwing an exception

BartoszKlonowski committed:
Fix: copy-paste error  'type' -> 'constructor'

BartoszKlonowski committed:
Align Mono and CoreCLR with nameof for GetConstructor

BartoszKlonowski committed:
Use type for argument exception and test against type parameter

eerhardt commented:
Tests are still failing:

```
System.Reflection.Emit.Tests.TypeBuilderGetConstructor.GetConstructor_TypeNotGeneric_ThrowsArgumentException [FAIL]
      Assert.Equal() Failure
                ? (pos 0)
      Expected: constructor
      Actual:   type
                ? (pos 0)
      Stack Trace:
        /_/src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs(123,0): at System.AssertExtensions.Throws[T](String expectedParamName, Func`1 testCode)
        /_/src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderGetConstructor.cs(77,0): at 
```

Are you able to run them locally? That would give you a faster turn around to get the CI green.

BartoszKlonowski commented:
@eerhardt I'm afraid I won't be able to finish this, sorry. I really appreciate your help and support, but analysing this and fixing the issue with tests would require much more time than I currently have...

steveharter commented:
@eerhardt if you want to take over this PR, please do so ASAP otherwise I recommend closing. Thanks

eerhardt commented:
I'll get to this PR once I am finished with my 6.0 deliverables.

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into fix-typebuilder-checks-order

eerhardt committed:
Get tests passing on both coreclr and mono.

Ensure checks are consistent across GetConstructor, GetField, and GetMethod for both coreclr and mono.

buyaa-n commented on src/mono/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.Mono.cs:
```diff
@@ -1898,20 +1897,19 @@ private static bool IsValidGetMethodType(Type type)
         {
             if (!IsValidGetMethodType(type))
-                throw new ArgumentException("type is not TypeBuilder but " + type.GetType(), nameof(type));
+                throw new ArgumentException("type must contain a TypeBuilder as a generic argument.", nameof(type));
```
NIT: a bit inconsistency with a period, looks they don't put a period on not capitalized exception messages

danmoseley replied:
shouldn't this be a in a resx?

buyaa-n replied:
Yeah, not sure why there are lots of string literals used in this file instead of using resources

lambdageek replied:
> Yeah, not sure why there are lots of string literals used in this file instead of using resources

Historical reasons. When this code was initially imported from mono/mono it made sense to keep the implementations similar for each of patching them in sync.  The mono/mono code didn't use string resources.
At this point it's better to use the same string resources are CoreCLR, if possible.

danmoseley replied:
as an aside, it's my understanding that any replacement strings that aren't just numbers need quotes. This is because if we ever localize, the localizers can't make the sentence grammatical without knowing what the replacement string is.

so `"type is not TypeBuilder but " + type.GetType() + " instead."` would not be allowable but something like this would be : `"type is not TypeBuilder but '" + type.GetType() + "' instead."`

`'type' is not TypeBuilder.` would be OK, but not as helpful because it doesn't include the type.

eerhardt replied:
> shouldn't this be a in a resx?

Yes, according to the bug you logged last year https://github.com/dotnet/runtime/issues/34056 😄

eerhardt replied:
I've fixed the exception messages I touched to use the same resources as CoreCLR. I left the ones I didn't touch since they don't correspond directly to existing CoreCLR exception messages. Those can get fixed with #34056.

buyaa-n reviewed (approved):
LGTM, thanks!

eerhardt committed:
Fix mono exception messages to use SR resources.
